File: leetcode/0152-maxProduct.py
class Solution:
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        res = cur_max = cur_min = nums[0]
        for num in nums[1:]:
            cur_max, cur_min = max(cur_max * num, cur_min * num, num), min(cur_max * num, cur_min * num, num)
            res =  max(res, cur_max)
        return res

File: leetcode/0152-maxProduct.java
class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0)
            return 0;
        int res = nums[0], curMax = nums[0], curMin = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int oldCurMax = curMax;
            curMax = Math.max(nums[i], Math.max(curMax * nums[i], curMin * nums[i]));
            curMin = Math.min(nums[i], Math.min(oldCurMax * nums[i], curMin * nums[i]));
            res = Math.max(curMax, res);
        }
        return res;
    }
}
